kubelet: Don't reference the pod manager interface directly from components

smarterclayton opened the pull request:
As part of #116970, we want to switch a number of components that are driven by the pod worker to reference the pod worker for the "actual" state of pods, rather than consulting the pod manager (which carries the "desired" state and does not know about orphaned pods).

This refactors pod manager to not get passed around as an interface to all components, but instead is passed by shape to each component. It also streamlines the methods on pod manager so that it is easier to implement. After this PR merges, we will be able to implement a pod manager equivalent on pod worker and have volume manager, eviction manager, probe manager, and status manager reference the pod worker instead of pod manager directly.

No functional change is intended in the code, this is only refactors.  Extracted from #115342

/kind cleanup
/sig node

/assign @bobbypage @rphillips 

```release-note
NONE
```
```docs

```

smarterclayton committed:
kubelet: Remove unused mirrorPodFunc in eviction

Not referenced

smarterclayton committed:
kubelet: Convert IsMirrorOf to a function

Shrinks the PodManager interface by one method, no abstraction is
necessary here.

smarterclayton committed:
kubelet: Merge orphaned mirror pod names into GetPodsAndMirrorPods

There is only one caller and both sets of data are part of the
resync operation between kubelet's desired state and the actual
state of the pod workers. Reduces the size of the interface so
that it is easier to create another pod manager.

smarterclayton committed:
kubelet: Remove dispatchWork and inline calls to UpdatePod

The HandlePod* methods are all structurally similar, but accrued
subtle differences. In general the only point for Handle is to
process admission and to update the pod worker with the desired
state of the kubelet's config (so that pod worker can make it
the actual state).

Add a new GetPodAndMirrorPod() method that handles when the config
pod is ambiguous (pod or mirror pod) and inline the structure.
Add comments on questionable additions in the config methods for
future improvement.

Move the metric observation of container count closer to where
pods are actually started (in the pod worker). A future change
can likely move it to syncPod.

smarterclayton committed:
kubelet: Separate the MirrorClient from the PodManager

The two are not coupled except accidentally. Separate them and
update callsites. This will reduce the scope of PodManager interface
to make exposing the pod worker cleaner.

smarterclayton committed:
kubelet: Organize and document kubelet pod-related members

Clearly describe core pod related component responsibilities in
the kubelet members. Organize the PodManager interface for clarity.

smarterclayton committed:
kubelet: Reduce the interface pod.Manager consumers accept

Every component that uses a pod.Manager should use a stub interface
(like we do for podWorker) that explicitly describes what methods
they use. This will allow podWorker to implement the minimum set
of manager interfaces.

smarterclayton committed:
kubelet: Rename PodManager DeletePod to RemovePod

RemovePod is more consistent within the kubelet to be the opposite
of AddPod, and the pod is not being deleted just "removed" from
tracking.

bart0sh commented:
/triage accepted
/priority important-longterm

smarterclayton commented:
/retest

dims commented:
/assign @mrunalp @SergeyKanzhelev

rphillips commented on pkg/kubelet/kubelet.go:
```diff
@@ -2535,12 +2584,21 @@ func (kl *Kubelet) HandlePodUpdates(pods []*v1.Pod) {
+		pod, mirrorPod, wasMirror := kl.podManager.GetPodAndMirrorPod(pod)
+		if wasMirror {
+			if pod == nil {
+				klog.V(2).InfoS("Unable to find pod for mirror pod, skipping", "mirrorPod", klog.KObj(mirrorPod), "mirrorPodUID", mirrorPod.UID)
```
If the log lines included the pod.UID it would enable better debugging if the mirror pod can't be found. We could include this in a followup PR.

smarterclayton replied:
+1, my gap

smarterclayton replied:
Ok, this is slightly tricky.  If we get called here we don't know the pod UID, because `pod` was a mirror pod, and mirror pods don't have the source UID on them (that I know of).  Definitely a followup.

smarterclayton replied:
Yeah, we don't have any way to find the real UID, because all we have is:

```
metadata:
  annotations:
    kubernetes.io/config.hash: 063e897b129c1f6d595e9010a8ac70ea // the hash
    kubernetes.io/config.mirror: 063e897b129c1f6d595e9010a8ac70ea // the hash
    kubernetes.io/config.seen: "2023-05-08T17:32:16.730248098Z"
    kubernetes.io/config.source: file
```

rphillips commented:
/lgtm

We can take my comment as a followup PR.

sohankunkerkar commented on pkg/kubelet/pod/pod_manager.go:
```diff
@@ -80,7 +80,7 @@ type Manager interface {
 	// DeletePod deletes the given pod from the manager.  For mirror pods,
```
`s/DeletePod/RemovePod`

rphillips replied:
We could potentially fix this in the followup PR I mentioned above too.

smarterclayton commented:
Fixed the comments,

/hold cancel

Please rereview!

smarterclayton commented:
/retest

rphillips commented:
Thanks for the changes!

LGTM

@bobbypage ?

rphillips commented:
@smarterclayton What are your thoughts on backporting this patch to 1.27?

smarterclayton commented:
If this applies, it is safe to put in 1.27 and I would be ok with it (to facilitate backports of other fixes we find)

bobbypage commented on pkg/kubelet/kubelet.go:
```diff
@@ -935,34 +935,107 @@ type Kubelet struct {
+	// resources must consult the podWorkers.
+	//
+	// Once a pod has been accepted by the pod workers, no other pod with that same UID (and
+	// fullName, for static pods) will be started until the first pod has fully terminated and
```
nit: Does it make sense to explain what `fullName` means here? i.e. fullName is pod name + namespace

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -66,11 +65,26 @@ type versionedPodStatus struct {
 }
 
+// podManager abstracts access to information about pods on the Kubelet.
+type podManager interface {
```
nit: It may make sense to document the interface here similar to the one in volume manager, i.e. something like:

```
// podManager is the subset of methods the status manager manager needs to observe the actual state of the kubelet
```

(and then the actual godoc for all the methods can be on the main pod manager interface). Otherwise, it will be duplicated between the "subset" of the interface here and the actual interface in `pod_manager.go`. As a result, I'm guessing will be tricky to be keep in sync between them.

smarterclayton replied:
Agreed

bobbypage commented:
Thank you for refactoring, these changes of minimizing the interface and dependencies will make it much easier to later add a new pod manager interface that will read from the "actual" state as compared to desired state now. I left a few small comments/questions.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -935,34 +935,107 @@ type Kubelet struct {
-	// podWorkers handle syncing Pods in response to events.
+	// podManager stores the desired set of admitted pods and mirror pods that the kubelet should be
+	// running. The actual set of running pods is stored on the podWorkers. The manager is populated
+	// by the kubelet config loops and may be consulted by other components that need to see the set
```
That's a good point, and worth adding.

smarterclayton commented on pkg/kubelet/pod/pod_manager.go:
```diff
@@ -324,3 +328,15 @@ func (pm *basicManager) GetPodByMirrorPod(mirrorPod *v1.Pod) (*v1.Pod, bool) {
+	if kubetypes.IsMirrorPod(aPod) {
+		return pm.podByFullName[fullName], aPod, true
+	}
+	return aPod, pm.mirrorPodByFullName[fullName], false
```
So pods can be regular or mirror.  You pass an unknown pod type here, and you get back either pod, mirror pod, or both.  From a Kube state diagram:

```
 +->  API --- (spec+mirror) -->-- config loop -->-- pod manager -->-- admission -->-- pod worker
 |                            |                                                           |
 |   File -- (spec) ----------+                                                           |
 |                                                                                        |
 +-------<--------------<------------------ status manager <------------------------------+
```

Mirror pods feed in to API, which means config loop has both, and then before admission we have to separate out the mirror pods and the spec pods.

So the method here takes either, and finds the ones we know about, and only "no pod" is stops us from going to pod manager.

bobbypage replied:
Got it, that makes sense. 

I was referring that we already check if `aPod` is a mirror pod, so if it is **not** a mirror pod, I would assume then that `pm.mirrorPodByFullName[fullName]` would be empty. As a result, I was thinking it might be clearer to do `return aPod, nil, false`

smarterclayton replied:
Hrm, we want to return the mirror pod in that case though, so that it can be passed in to the pod worker.

There's an argument that mirror pod management is a sub responsibility of pod worker (sync methods), not pod manager.  We still need to propagate the config loop items into pod worker.  I think the argument is that the "config -> actual" transition is really a set of controller methods that react to incoming events and notify the right components (mostly pod worker, but maybe others eventually), with a resync method that periodically verifies no delta was missed.  That might be moving all the HandlePod* methods out of kubelet and into a type that calls back to kubelet for specific stuff, but mostly deals with pod worker and a few subsystems.  Maybe not high priority though.

smarterclayton replied:
We need the mirror pod from this method in all cases.  Maybe I misunderstand the question.

bobbypage replied:
ah, my mistake, I read this over again and it makes sense. The key point is that when the argument `aPod` is passed in as not a mirror pod (i.e. regular pod), this regular pod may or may not have a mirror pod, so ` pm.mirrorPodByFullName[fullName]` may or may not be nil depending if a mirror pod exists for it. 

So as you mentioned, we need to return mirror pod in all cases.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2535,12 +2523,21 @@ func (kl *Kubelet) HandlePodUpdates(pods []*v1.Pod) {
+
+		kl.podWorkers.UpdatePod(UpdatePodOptions{
+			Pod:        pod,
+			MirrorPod:  mirrorPod,
```
Hrm, we could.  It also makes the code more variable.  Hrm - I can see the argument to making it different, but then each method is subtly different.

smarterclayton replied:
Would prefer to leave them the same for now.

bobbypage replied:
sgtm!

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2550,10 +2547,22 @@ func (kl *Kubelet) HandlePodRemoves(pods []*v1.Pod) {
+				UpdateType: kubetypes.SyncPodUpdate,
+				StartTime:  start,
+			})
 			continue
```
Good catch.  Let me go through the logic and think about whether this is subtly wrong.

smarterclayton replied:
We had a continue before - 

```
if kubetypes.IsMirrorPod(pod) {
  kl.handleMirrorPod(pod, start)
  continue
}
```

was the old code

bobbypage replied:
thanks for pointing it out, I misread the location of `continue`, it's not at the end of the for pod loop.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2571,13 +2580,38 @@ func (kl *Kubelet) HandlePodReconcile(pods []*v1.Pod) {
+				klog.V(2).InfoS("Unable to find pod for mirror pod, skipping", "mirrorPod", klog.KObj(mirrorPod), "mirrorPodUID", mirrorPod.UID)
+				continue
+			}
+			// We only expect to be called on regular pods
```
yes we should

smarterclayton replied:
Hrm, now I'm questioning why we have this continue.

smarterclayton replied:
The continue was wrong, I added two comments once I brought into memory why reconcile exists (pod ready added it in order to know when a 3rd party component updates ready).  It's probably too specific (reconcile should just go ahead and poke the worker) but not going to change it here.  Asked https://github.com/kubernetes/kubernetes/pull/64344/files#r1191677121 to verify

bobbypage replied:
That makes sense, thanks for following up about it, agreed we can defer it to later as set of future cleanups.

bobbypage replied:
That makes sense, thanks for following up about it, agreed we can defer it to later as set of future cleanups.

smarterclayton commented on pkg/kubelet/kubelet.go:
```diff
@@ -2591,8 +2625,21 @@ func (kl *Kubelet) HandlePodReconcile(pods []*v1.Pod) {
+				klog.V(2).InfoS("Unable to find pod for mirror pod, skipping", "mirrorPod", klog.KObj(mirrorPod), "mirrorPodUID", mirrorPod.UID)
+				continue
+			}
+			// since we are syncing all pods, we don't have to sync mirrors
```
Before, if we got passed a mirror pod, we would send the mirror pod to UpdatePod, which is not correct.  pod workers should never receive a mirror pod in the `Pod` field.  HandlePodSyncs only accepts "real pods".  So this would be programmer error (even if it never happened).  I can add a klog.

bobbypage replied:
Got it, it makes sense we shouldn't sync mirror pods to `UpdatePod`. I was confused because I saw we _were_ doing that before, but it sounds like that was a bug and we shouldn't have been doing that in the first place.

bobbypage commented on pkg/kubelet/pod_workers.go:
```diff
@@ -1174,6 +1174,12 @@ func (p *podWorkers) startPodSync(podUID types.UID) (ctx context.Context, update
+	// TODO: We should probably move this into syncPod and add an execution count
+	// to the syncPod arguments, and this should be recorded on the first sync.
+	// Leaving it here complicates a particularly important loop.
+	metrics.ContainersPerPodCount.Observe(float64(len(update.Options.Pod.Spec.Containers)))
```
oh right, that makes sense - it means we shouldn't get here more than once in the first place since we return if `isStarted` is true.

smarterclayton commented:
Updated with all the things I could tweak (a few more comments in various places).  Still have the two outstanding questions back to David.

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -87,8 +86,18 @@ type manager struct {
-// that need to introspect status.
+// PodManager is the subset of methods the manager needs to observe the actual state of the kubelet.
+// See pkg/k8s.io/kubernetes/pkg/kubelet/pod.Manager for method godoc.
+type PodManager interface {
```
is there any reason this needs to be exported? Can it be `podManager`?

bobbypage commented on pkg/kubelet/status/status_manager.go:
```diff
@@ -87,8 +86,18 @@ type manager struct {
-// that need to introspect status.
+// PodManager is the subset of methods the manager needs to observe the actual state of the kubelet.
+// See pkg/k8s.io/kubernetes/pkg/kubelet/pod.Manager for method godoc.
+type PodManager interface {
```
Is there any reason this needs to be exported? Same question in volume manager.

smarterclayton replied:
on the principle that the interface is being used in a public method.  We have more cases where we export the interface than not, and ALSO you don't see public godoc for private methods, so leaned slightly towards making them all public.  In *general* go style discourages private types in public methods, but it's not hard and fast, and Kube doesn't consistently enforce it.

bobbypage replied:
I see, that makes sense - if it's being used in a public method then it makes sense to export the interface as well.

bobbypage commented:
Thanks for the updates, LGTM! I replied in the thread about the two questions. Just a small comment about why we need to export the interfaces.

It also looks like the mocks need to be regenerated by the `pull-kubernetes-verify`, job, i.e.

```
Mock files are out of date
diff --git a/pkg/kubelet/status/testing/mock_pod_status_provider.go b/pkg/kubelet/status/testing/mock_pod_status_provider.go
...
Please run 'hack/update-mocks.sh'
```

smarterclayton committed:
test: Improve debug output of init container tests

When certain status conditions are not expected, we need to see
the nested objects, but %#v doesn't handle pointers well. Output
as simple encoded JSON.

smarterclayton commented:
https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/117371/pull-kubernetes-conformance-kind-ga-only-parallel/1656769645325062144 worries me, adding debug output to dig in.

smarterclayton commented:
/test pull-kubernetes-conformance-kind-ga-only-parallel

rphillips commented:
David LGTMed. LGTM as well.

/lgtm

bobbypage commented:
sounds good, lgtm as well!

/lgtm

smarterclayton commented:
https://storage.googleapis.com/k8s-triage/index.html?pr=1&text=container%20should%20have%20reason%20PodInitializing looks like it triggered, debugging
